### ISSUE TITLE
Consider changing the behavior of / route when nesting

### ISSUE THREAD

**Issue by jplatte** [tokio-rs/axum#2659]
Currently, given a router

```rust
let inner = Router::new().route("/", method_router);
```

and an outer router

- `Router::new().nest("/foo", inner)` => only the route `/foo` exists
- `Router::new().nest("/foo/", inner)` => only the route `/foo/` exists

It seems like people (me included) generally prefer the first form, but oftentimes expect different behavior (see #2267). We should (re)consider what behavior(s) are reasonable for the upcoming release. Also worth checking: What happens with `nest_service`?

**Comment by jplatte** [comment]
cc @yanns @mladedav @ibraheemdev do any of you have ideas / opinions on this?

**Comment by mladedav** [comment]
I've also sometimes guessed wrong how the nesting with root handlers works. One option would be to simply concatenate, but there should be rules to disallow things like  this:
```rust
Router::new().nest("/foo", Router::new().route("bar", todo)); // matches `/foobar` ??
```

What I imagine could work is that axum would require:
- paths in `nest`
  - must start with a slash
  - must not end with a slash
- paths in `route`
  - must be empty or start with a slash

For example:
```rust
Router::new().nest("", Router::new().route("/foo", todo)); // error - nest must start with a slash
Router::new().nest("/", Router::new().route("/foo", todo)); // error - nest must not ends with a slash
Router::new().nest("/foo", Router::new().route("", todo)); // matches `/foo`
Router::new().nest("/foo", Router::new().route("/", todo)); //matches `/foo/`
Router::new().nest("/foo", Router::new().route("/bar", todo)); //matches `/foo/bar`
Router::new().nest("/foo", Router::new().route("/bar/", todo)); //matches `/foo/bar/`
Router::new().nest("/foo/", Router::new().route("/bar", todo)); // error - nest ends with a slash
Router::new().nest("/foo", Router::new().route("bar", todo)); // error - route doesn't start with a slash
```

Currently, axum only requires all paths to start with a slash as far as I know.

This would allow setting routes with or without trailing slashes and would even prevent users from building the same router multiple ways (`nest` with or without a trailing slash is identical except for handling the root as far as I know).

The thing I dislike about this the most is that
```
let router = Router::new().route("/", root).route("", empty);
```
would behave differently based on whether it would be nested or not. On top level the two routes are equivalent (so we should maybe even reject starting the server due to conflicting routes) but in a nested router they are different.

This is because [RFC 3386](https://www.rfc-editor.org/rfc/rfc3986.html#section-6.2.3) mentions that `example.com` and `example.com/` are the same, but later states that `http://example.com/data` should be considered the same as `http://example.com/data/` only if the server hints that they are the same for example by redirects between them.

One way out of this would be to either let the user provide the `""` handler only when they call `nest` but this could be elsewhere than the rest of the inner router definition.

Another would be to make the empty route even more special in a way that it returns different type, e.g. `NestableRouter` which can be used only for `Nest` and not for starting the app. I think this would be preferred if possible.

**Comment by yanns** [comment]
> cc @yanns @mladedav @ibraheemdev do any of you have ideas / opinions on this?

I have never used `nest` myself, having the feeling (maybe wrong) that it can be dangerous with lots of subtleties.
So I would not comment on this.

**Comment by timhughes** [comment]
This was discussed back in 2022 https://github.com/tokio-rs/axum/issues/714 https://github.com/tokio-rs/axum/issues/1118 but I cannot find where the decisions were made

You can use NormalizePath 

```rust
use axum::{
    extract::Request, routing::get, routing::get_service,
    Router, ServiceExt,
};
use tokio::net::TcpListener;
use tower_http::{
    normalize_path::NormalizePathLayer,
    services::{ServeDir, ServeFile},
};
use tower_layer::Layer;

#[tokio::main]
async fn main() {
    fn app_router() -> Router {
        Router::new()
            .route("/", get(|| async { "hi" }))
    }

    let serve_dir =
        get_service(ServeDir::new("static").not_found_service(ServeFile::new("static/index.html")));

    let app = Router::new()
        .nest("/api", app_router())
        .nest_service("/", serve_dir);

    let app = NormalizePathLayer::trim_trailing_slash().layer(app);
    let app = ServiceExt::<Request>::into_make_service(app);

    let listener = TcpListener::bind("127.0.0.1:3000").await.unwrap();
    axum::serve(listener, app)
        .await
        .unwrap();
}
```

**Comment by timhughes** [comment]
there is also some strange behaviours around `/path` vs `/path/` when you have a `.fallback()`

I have made a gist with the details https://gist.github.com/timhughes/745a07746f96c64586682d78829b840b

**Comment by jplatte** [comment]
@mladedav I have (re-)reviewed your proposal above and I quite like it. I think I previously assumed that `route("", ...)` wasn't even supported, but if it indeed is, let's take advantage of that and have exactly one way to do nested paths.

W.r.t. what to do when both an empty route and a `/` route are defined, I think it's not really possible to define when a router is the "root" so my preference would be to do nothing. I'm guessing the empty route would never actually be used since hyper probably normalizes the URL. That seems alright to me, people will figure it out fast enough if they ever think to register the empty route.

Also, I've moved this to the 0.9 milestone. I think we've got enough breaking changes in 0.8 that people need to adjust to. Lmk if you disagree.

**Comment by mladedav** [comment]
> I think I previously assumed that `route("", ...)` wasn't even supported

It isn't in the current version of axum, just in the proposal. Not sure which one you meant here.

> W.r.t. what to do when both an empty route and a / route are defined, I think it's not really possible to define when a router is the "root" so my preference would be to do nothing. I'm guessing the empty route would never actually be used since hyper probably normalizes the URL.

That's not just hyper, but the whole of HTTP I think. There is no empty path, just `/`. But there are `/foo` and `/foo/` and these can be different. So if we let people define handler for the empty path (and the router is not nested), they will never hit that handler. We can also panic on `serve` in that case.

**Comment by xaberus** [comment]
> Also worth checking: What happens with `nest_service`?

Well, `nest_service` routes both paths to `/` of the nested router. At the same time, `nest` routes only the _exact_ match to `/`. I hope there is a way to make both methods consistent...

To reproduce, consider the following code:
```rust
use axum::{
    Router,
    body::Body,
    extract::Request,
    http::StatusCode,
    response::{IntoResponse, Response},
    routing::get,
};
use futures::executor::block_on;
use tower_service::Service;

async fn handler() -> Response {
    StatusCode::OK.into_response()
}

fn main() {
    let inner = Router::new()
        .route("/", get(handler))
        .route("/bar", get(handler));
    let mut nest: Router<()> = Router::new().nest("/foo", inner.clone());
    let mut nest_service: Router<()> = Router::new().nest_service("/foo", inner);

    fn test(app: &mut Router, uri: &str) {
        let req = Request::builder()
            .uri(format!("https://example{}", uri))
            .body(Body::empty())
            .unwrap();
        let rsp = block_on(app.call(req)).unwrap();
        println!("  {} -> {}", uri, rsp.status());
    }

    println!("nest():");
    test(&mut nest, "/foo");
    test(&mut nest, "/foo/");
    test(&mut nest, "/foo/bar");
    test(&mut nest, "/foo/baz");
    println!("nest_service():");
    test(&mut nest_service, "/foo");
    test(&mut nest_service, "/foo/");
    test(&mut nest_service, "/foo/bar");
    test(&mut nest_service, "/foo/baz");
}
```
and its output:
```
nest():
  /foo -> 200 OK
  /foo/ -> 404 Not Found
  /foo/bar -> 200 OK
  /foo/baz -> 404 Not Found
nest_service():
  /foo -> 200 OK
  /foo/ -> 200 OK
  /foo/bar -> 200 OK
  /foo/baz -> 404 Not Found
```

**Comment by bowenxuuu** [comment]
Any progress?

**Comment by ficolas2** [comment]
I would like to give this a go

**Comment by mladedav** [comment]
@davidpdrsn Do you maybe have an opinion on this/#3812?

Alternatives I see are not allowing empty route paths and adding such handlers in the `nest` call
```rust
Router::new()
  .route("/foo", get(async{}))
  .nest("/foo", Router::new())
  .nest_with_root_handler("/bar", Router::new(), get(async {}))
```

Or we could tell the `nest` call where to register the root route with something like
```rust
Router::new()
    .nest_with_root_behavior("/foo", Router::new().route("/", get(async {"root"})), NestRootBehavior::WithoutSlash)
```

This doesn't allow us to have different handlers for `""` and `"/"` in the same nested router though. I don't think that's an issue though. People can still add the other route on the top level router.

And finally having separate types for `Router` which doesn't allow empty paths and `NestableRouter` which allows it but doesn't implement `Service` so that it can't be passed to `axum::serve`. But I worry this could be unnecessarily complicated.

---

I think I'm personally leaning towards not allowing the empty paths since that looks weird and could be a footgun for top level routers. Instead having users explicitly choose nest behavior (with the default being nesting without trailing slash for `nest`) seems to me like the best option that still allows us to reject trailing slashes in all nests.

**Comment by jlizen** [comment]
@mladedav I tend to agree with the idea of moving root behavior handling to the nest. I think the ergonomics would be better with builders though. Something along the lines of:

```rust
  use axum::routing::nested;

  Router::new()
      // common case unchanged
      .nest("/api", api)
      // opt-in configuration, chained
      .nest("/foo", nested(inner).root(get(root_handler)))
      .nest("/bar", nested(other).root_from_slash_route())
```

With the implementation being something like:
```rust
  pub struct Nested<S> {
      router: Router<S>,
      root: RootBehavior<S>, // None | Handler(MethodRouter<S>) | InnerSlashRoute
  }

  impl<S> Nested<S> {
      pub fn root(mut self, handler: MethodRouter<S>) -> Self { ... }
      pub fn root_from_slash_route(mut self) -> Self { ... }
  }
```

WDYT?

**Comment by mladedav** [comment]
This looks to me more like the "servable and nestable routers are a different type" but it seems much simpler than I originally thought it would.

One thing I notice is that if we want to use `nest` both for `Router` and `Nested`, we'll need to accept something like `impl Into<Nested>` and implement that for `Router` which should be straightforward.

Also we could maybe have `root` and `root_from_slash_route` directly on `Router`. For better or worse that would allow users to call either of those methods just once. But it would get rid of the `nested` calls.

With that it could look something like:

```rust
Router::new()
  .nest("/foo", Router::new())
  .nest("/bar", Router::new().root(get(async{})));
```

---

I think this definitely better than allowing `route` with empty route because `root` does effectively the same thing but we can prevent users from passing a router with that into `serve`.

So yeah, having `Nested` like this looks good to me.

**Comment by jlizen** [comment]
Yeah, that looks much nicer with the `root()` and `root_from_slash_route` as methods on the router.  I'm on board with that plan.
